src/basilisp/test.lpy

```diff
@@ -284,15 +438,91 @@
+  (-> report
+      (report-test-begin test-var)
+      (report-test-end test-var (execute* fixtures test-var))))
+
+(defn- report-namespace
+  [report ns test-vars]
+  (let [report       (report-namespace-begin report ns)
+        once-fixture (-> ns meta ::once-fixtures join-fixtures)
+        each-fixture (-> ns meta ::each-fixtures join-fixtures)]
+    (try
+      (report-namespace-end (with-fixtures [once-fixture]
+                              (reduce (fn [report test-var]
+                                        (if (continue? report)
+                                          (report-test report each-fixture
+                                                       test-var)
+                                          (reduced report)))
+                                      report
+                                      test-vars))
+                            ns
+                            nil)
+      (catch Exception e
+        (report-namespace-end report
+                              ns
+                              [{:test-name    nil
+                                :test-section nil
+                                :ns           ns
+                                :var          nil
+                                :expr         nil
+                                :line         nil
+                                :type         :error
+                                :expected     nil
+                                :actual       e
+                                :message
+                                (str "Exception in namespace fixture: "
+                                     (python/repr e))}])))))
+
+(defn collect-tests
+  "Get all tests from the given vars and/or namespaces. If none are
+  given then default to the tests current namespace. Return a mapping
+  of namespaces to test vars."
+  [vars-or-namespaces]
+  (->> (or (seq vars-or-namespaces)
+           [*ns*])
+       (mapcat #(if (var? %) (list %) (-> % ns-interns vals)))
+       (filter (comp ::test meta))
+       (group-by namespace)))
+
+(defn compile-report
+  "Build a report by running all tests from the given vars and/or
+  namespaces. If none are given then default to the tests current
+  namespace."
+  [report vars-or-namespaces]
+  (report-end
+   (reduce (fn [report [ns test-vars]]
+             (if (continue? report)
+               (report-namespace report ns test-vars)
+               (reduced report)))
+           (report-begin report)
+           (collect-tests vars-or-namespaces))))
+
+(defn run-tests
+  "Run all tests from the given vars and/or namespaces. If none are
+  given then default to the tests current namespace. Print failures to
+  standard output."
+  [& vars-or-namespaces]
+  (into {} (compile-report (->SimpleReport 0 0 0 0) vars-or-namespaces)))
+
+(defn run-all-tests
+  "Run all loaded tests. Print failures to standard output. Optionally
+  filter tests by regex ``re``."
+  ([]
+   (apply run-tests (all-ns)))
```

Comment:
I don't think `all-ns` works like you think unless you've already somehow required the associated test namespaces.

Reply:
> This is how it works in `clojure.test`.

Indeed, but I think from spot checking that Basilisp is far more conservative with loading namespaces than Clojure. 

I haven't developed on a large Clojure project in some time, so I can't confirm with a larger codebase but even starting `clj` locally I see many more namespaces listed than in Basilisp:

```
user=> (map ns-name (all-ns))
(clojure.edn clojure.spec.gen.alpha clojure.core.server clojure.java.basis clojure.java.io clojure.java.shell clojure.core.protocols clojure.string clojure.java.browse clojure.pprint clojure.java.basis.impl clojure.uuid clojure.core clojure.main user clojure.java.javadoc clojure.repl.deps clojure.tools.deps.interop clojure.walk clojure.repl clojure.instant clojure.java.process clojure.spec.alpha)
```

I'm not sure if Clojure (or `clj` or Leiningen) automatically load every namespace in the classpath at startup, but that is not likely something I would want to do due to long compile times.

Reply:
Got it. I suppose this won't be an issue, then.